### PR TITLE
AO3-3846 Make draft chapters display on chapter rearrange page

### PR DESCRIPTION
**Pull request by melinath**
# Pull Request Checklist

* [x] Have you read [How to write the perfect pull request](https://github.com/blog/1943-how-to-write-the-perfect-pull-request)?
* [x] Have you read through the [contributor guidelines](https://github.com/otwcode/otwarchive/blob/master/CONTRIBUTING.md)?
* [x] Have you added tests for any changed functionality?
* [x] Have you added the [JIRA](https://otwarchive.atlassian.net) issue number as the *first* thing in your pull request title (eg: `AO3-1234 Fix thing`)
* [x] Have you updated or commented on the JIRA issue with the information below?

## Issue

https://otwarchive.atlassian.net/browse/AO3-3846

## Purpose

Makes it so that draft chapters are displayed on the chapter rearranging page rather than excluded. This allows them to be rearranged in the same way as other chapters.

## Testing Instructions

1. Create a multi-chapter work.
2. Add a draft chapter to the work.
3. Visit the work's Edit page.
4. Visit "Manage Chapters".
5. Verify that the draft chapter appears on the page.
6. Move draft chapter and save.
7. Verify that draft chapter number was updated.

## Credit

*What name do you want us to use to credit your work in the [Archive of Our Own's Release Notes](https://archiveofourown.org/admin_posts?tag=1)?*

Stephen Lewis

*What pronouns do you prefer (she/her, he/him, zie/hir etc)?*

he/him

### CODE REVIEW

**Comment by melinath** [comment]
oh no! That makes total sense, sorry!

**Comment by sarken** [comment]
(First, I want to say that know it's been quite a while, so I understand if you're no longer interested in working on this, @melinath. If you are, there's no rush -- I'm still not sure when we'll be merging this, since we've run into some confusion around the issue it's dependent on. However, I did want to record some notes here so we didn't get surprised later on.)

I just tried this code to see if it would help me reproduce AO3-4313, and I'm not 100% sure reordering draft chapters works. 

While the manage chapters page does _display_ the draft chapter, when I use drag and drop to reorder them, the chapters stay in the same order after I press "Update Positions." 

If I try reordering the chapters by entering their new numbers in the text fields, I get a 500 error when `reorder_list` is called on line 169 here:

https://github.com/otwcode/otwarchive/blob/40fb60907f998e86424c097b4a4edfca1ca3e0f5/app/controllers/chapters_controller.rb#L167-L181

The error (`undefined method 'update_attribute' for nil:NilClass`) is on line 12 of the method:

https://github.com/otwcode/otwarchive/blob/40fb60907f998e86424c097b4a4edfca1ca3e0f5/lib/sortable_list.rb#L2-L13

It might just be me, but before we merge this, it would be a good idea to expand the tests to make sure reordering draft chapters works, as opposed to only checking if they're displayed.

**Comment by melinath** [comment]
Just to have this recorded - I am definitely still down to look at this - thanks for catching the bug. If it's not a priority I will probably hold off on it for a bit longer, especially since it's still blocked.

**Comment by sarken** [comment]
Thanks for checking in! This isn't urgent at all, so you're welcome to take as long as you need. We'll be here when you're ready.

**Comment by melinath** [comment]
Okay, so I've looked into this a bit and it looks like technically the reordering by dragging does work even with the draft works - it's just that it happens _immediately_. If you press "Update Positions", that makes a separate request using the explicit number fields (but with the order values set to empty strings) which seems to preserve the ordering of the chapters (but currently shifts all the draft chapters to the end, probably in an arbitrary order?) 

I should be able to update the endpoint to treat empty (and non-empty) values the same for draft and non-draft chapters when you press "update positions".

File: features/works/chapter_edit.feature
Comment (sarken) [inline features/works/chapter_edit.feature:156]: Let's check for "Chapter 5 (Draft)" to make sure line 27 of chapters/manage.html.erb is working, too.

**Review by sarken — COMMENTED** [review]
It looks good to me! I just had one tiny suggestion for the feature test.

**Review by sarken — APPROVED** [review]

**Comment by redsummernight** [comment]
Removing the "deferred" label now that we have #3985 for the blocking issue.

**Comment by sarken** [comment]
It looks like there's a merge conflict on this now. If you have a few minutes to resolve it, that would be great! Or you can let us know if you'd rather us deal with it.

**Review by sarken — COMMENTED** [review]
Thanks for updating this! I just have one small comment about a test description.

File: spec/controllers/chapters_controller_spec.rb
Comment (melinath) [reply spec/controllers/chapters_controller_spec.rb]: indeed! fixed in 1b14069d0

**Pull request merged**
